apps/speed.c

```diff
@@ -2800,7 +2800,9 @@ int speed_main(int argc, char **argv)
         for (i = 0; i < loopargs_len; i++)
             ASYNC_WAIT_CTX_free(loopargs[i].wait_ctx);
 
-        ASYNC_cleanup_thread();
+        if (loopargs_len != 0) {
```

Comment:
I don't like a no-op ASYNC_cleanup_thread() because it introduces a hidden state to the init code which doesn't give much advantage. The abstraction would be incomplete and wouldn't come close to a "with" statement in other languages.

ASYNC_init_thread() must still return a value to see how to progress with the code and clean-up must still be explicitly called. The only difference would be for speed.c in that clean-up is no longer guarded by an if.
For other constructs, different code arrangements may not require this conditional at all yet an explicit call to ASYNC_cleanup_thread() must still be made regardless of what it does.